Penteract/Assets/Scripts/CameraController.cpp

```diff
@@ -1,18 +1,69 @@
 #include "CameraController.h"
 
+#include "PlayerController.h"
+#include "Components/ComponentTransform.h"
+
+#define PI 3.14159
+
+
 EXPOSE_MEMBERS(CameraController) {
-    // Add members here to expose them to the engine. Example:
-    // MEMBER(MemberType::BOOL, exampleMember1),
-    // MEMBER(MemberType::PREFAB_RESOURCE_UID, exampleMember2),
-    // MEMBER(MemberType::GAME_OBJECT_UID, exampleMember3)
+	MEMBER(MemberType::FLOAT, cameraOffsetZ),
+	MEMBER(MemberType::FLOAT, cameraOffsetY),
+	MEMBER(MemberType::FLOAT, cameraOffsetX),
+	MEMBER(MemberType::FLOAT, smoothCameraSpeed),
+	MEMBER(MemberType::BOOL, useSmoothCamera),
+	MEMBER(MemberType::GAME_OBJECT_UID, playerControllerObjUID),
+	MEMBER(MemberType::FLOAT, shakeTotalTime),
+	MEMBER(MemberType::FLOAT, shakeTimer),
+	MEMBER(MemberType::FLOAT, shakeMultiplier)
+
 };
 
 GENERATE_BODY_IMPL(CameraController);
 
 void CameraController::Start() {
-	
+	transform = GetOwner().GetComponent<ComponentTransform>();
+	GameObject* playerControllerObj = GameplaySystems::GetGameObject(playerControllerObjUID);
+	if (playerControllerObj) {
+		playerController = GET_SCRIPT(playerControllerObj, PlayerController);
+	}
 }
 
 void CameraController::Update() {
-	
+
+	if (playerController == nullptr || transform == nullptr) return;
+	float3 playerGlobalPos = playerController->playerFang.playerMainTransform->GetGlobalPosition();
+
+
+	float3 desiredPosition = playerGlobalPos + float3(cameraOffsetX, cameraOffsetY, cameraOffsetZ);
+	float3 smoothedPosition = desiredPosition;
+
+	if (useSmoothCamera) {
+		smoothedPosition = float3::Lerp(transform->GetGlobalPosition(), desiredPosition, smoothCameraSpeed * Time::GetDeltaTime());
+	}
+
+	if (shakeTimer > 0) {
+		float2 shakeDir = GetRandomPosInUnitaryCircle(float2(0, 0));
+		transform->SetGlobalPosition(smoothedPosition + transform->GetRight() * shakeDir.x * shakeMultiplier + transform->GetUp() * shakeDir.y * shakeMultiplier);
+		shakeTimer -= Time::GetDeltaTime();
+	} else {
+		transform->SetGlobalPosition(smoothedPosition);
+	}
+}
+
+void CameraController::StartShake() {
+	shakeTimer = shakeTotalTime;
+}
+
+
+float2 CameraController::GetRandomPosInUnitaryCircle(float2 center) {
+	float random = (static_cast<float>(rand() % 101)) / 100.0f;
+	float random2 = (static_cast<float>(rand() % 101)) / 100.0f;
+
+	float r = sqrt(random);
+	float theta = random2 * 2 * PI;
+	float x = center.x + r * cos(theta);
+	float y = center.y + r * sin(theta);
```

Comment:
If GetRandomPosInUnitaryCircle always receive float2(0,0) as parameter, you could remove it.
If it's going to be used with a different center in the future, it's fine